pkg/serviceaccount/claims.go

```diff
@@ -162,7 +190,10 @@ func (v *validator) Validate(ctx context.Context, _ string, public *jwt.Claims,
 		return nil, fmt.Errorf("service account UID (%s) does not match claim (%s)", serviceAccount.UID, saref.UID)
 	}
 
-	if secref != nil {
```

Comment:
I actually prefer the old structure that handles each reference individually and doesn't assume they are mutually exclusive:

```go
if secref != nil {
  ... validate secret ref
}

var podName, podUID string
if podref != nil {
  ... validate pod ref ...
}

var nodeName, nodeUID string
if noderef != nil {
  switch {
    case podref != nil && utilfeature.DefaultFeatureGate.Enabled(features.ServiceAccountTokenPodNodeInfo):
      // for pod-bound tokens, just extract the node claims
      nodeName = noderef.Name
      nodeUID = noderef.UID

    case podref == nil && utilfeature.DefaultFeatureGate.Enabled(features.ServiceAccountTokenNodeBindingValidation) {
      // for node-bound tokens, extract and validate the node claims
  		node, err := v.getter.GetNode(noderef.Name)
      ...
		  nodeName = noderef.Name
		  nodeUID = noderef.UID

    case podref == nil:
			klog.V(4).Infof("ServiceAccount token is bound to a Node object, but the node bound token validation feature is disabled")
			return nil, fmt.Errorf("token is bound to a Node object but the %s feature gate is disabled", features.ServiceAccountTokenNodeBindingValidation)
  }
}
```

This means if we ever have a way to create tokens that add secret claims AND pod claims AND node claims, the validating code still does the right thing

Reply:
I've made this change :)